xarray/core/dataset.py

```diff
@@ -8943,21 +8943,21 @@ def groupby(
         Dataset.resample
         DataArray.resample
         """
-        from xarray.core.groupby import DatasetGroupBy
-
-        # While we don't generally check the type of every arg, passing
-        # multiple dimensions as multiple arguments is common enough, and the
-        # consequences hidden enough (strings evaluate as true) to warrant
-        # checking here.
-        # A future version could make squeeze kwarg only, but would face
-        # backward-compat issues.
-        if not isinstance(squeeze, bool):
-            raise TypeError(
-                f"`squeeze` must be True or False, but {squeeze} was supplied"
-            )
+        from xarray.core.groupby import (
```

Comment:
There is some duplication in dataarray, dataset here introduced by @headtr1ck in #6702. Is there a way to undo that and just have a simple entry in `common.DataWithCoords`?

Reply:
To keep import times as low as possible the groupby code is only imported when it is used, which was not possible with the previous implementation.

Also it is not really possible to type it correctly in DataWithCoords, because you don't know which of the two groupby objects you get.

Reply:
Maybe some repeated code can be factorized out, i haven't checked your changes yet...

Reply:
Ok thanks. Those are good points. I think the repetition isn't too bad, just a little painful when iterating on the interface